doc/maxwell.8
.TH MAXWELL 8 2016-07-30 Linux 

.SH NAME
Maxwell \- Daemon for gathering entropy from a timer and feeding it to
.BR random (4).

.SH SYNOPSIS.
.ll +8
.B maxwell
.RB [\-s ]
.B maxwell
[\-p 
.I number
]
[\-f 
.I number
]

.SH DESCRIPTION

.I maxwell
collects randomness from the small variations in a system
timer, distills it into a concentrated form, and sends it to
.BR random (4).
The amount of output is a few K bits per second,
and the quality
is intended to be very high. Both volume and quality should be
adequate for most applications.

Non-root users can run the program only with the -s (stdout)
or -<digit> options, which send test data to standard output.

There are a number of other ways to feed entropy to
.BR random (4).
The advantage of
.I maxwell
is that it is small, simple and only minimally
hardware-dependent.
The other methods also have advantages, and in
many cases one of them will be preferable to
this one. For discussion, see the design
paper listed in the references.

This man page is for maxwell version 2 (2016); the options have changed considerably from version 1 (2012).

.SH OPTIONS

Two options take no parameters; these are intended mainly for use in testing:

.TP 
.BI \-s
Send output to standard output instead of to /dev/random.

Other options let
you control the amount and type of output, and it can be
piped into test tools or sent to a file for later analysis.

A
.I non-root user cannot run the program without this option or
-<digit>. It will just output an error message and quit. With the option,
it can be run for testing. It can also be redirected into
/dev/random, though no entropy credit will be given.

.TP
.BI -<digit>
This runs the given number of loops, where without this option
at least three are always used. It also omits the extra mixing normally done outside
the loop. It sends output to standard out.

If
.I maxwell -1 | ent
shows >= three bits of entropy per byte, then
.I maxwell
is safe. That is 12 bits per loop and
.I maxwell
uses at least three loops per output.

.PP
Several options with numeric arguments allow changing the parameters
with which the program runs. The options may be omitted, in which case
default values are used.

The numeric arguments are required, and must be in the range 0-99.

.TP 
.BI \-p \ n
Paranoia, affects how much looping is done before
each output. The number of loops done is 2p + 3.
Each loop collects sixteen timer samples for entropy. 

Default is p = 0 which gives three loops.

.TP 
.BI \-f \ n
At startup, the program produces some output using
parameters chosen to quickly fill up the random(4)
pool. After that, it changes to more conservative
parameters for longer-term use. This parameter controls how much of the first fast
data is generated, in kilobits.

Default is 4, enough to completely fill the pool
if random(4) is compiled with its default parameters.

.TP 
.BI \-h \ n
Halt after n kilobits of output.

Default is for the program to continue indefinitely,
but limiting the output is useful for testing.

.PP
The above options may be combined at will.

.SH BUGS

The exact entropy generated by a timer call cannot be known, so it
is impossible to know exactly how much entropy the program feeds to
.BR random (4).

The program uses a loop to collect entropy. Each pass makes 16 timer
calls and gets 16 samples modulo 31. With the default settings, there
are three passes for each 32-bit chunk of output. This is safe if 16
timer calls give 11 or more bits of entropy.

There is considerable evidence that available entropy is much more
than this minium requirement. See the design paper.

.SH TESTING

In addition to the -s option described above, there is a test
program called test.sam.c in the source directory. It is designed
to have its output piped into ent(1), an entropy testing program.

This tests at a lower level than maxwell -s; it samples data in
the same way maxwell does, but does no mixing at all, just packs
each sample into a byte and sends it out.

See the design paper and comments in the test program
source for details.

On any unusual system, these tests should be run
before
.BR maxwell ()
is deployed.

.SH SEE ALSO

.BR random (4)
.BR ent (1)

.SH References

RFC 4086
.I Randomness Requirements for Security
is the standard reference on randomness issues.
http://www.ietf.org/rfc/rfc4086.txt

.I Analysis of inherent randomness of the Linux kernel
by Nicholas McGuire, Peter Okech & Georg Schiesser
discusses timer entropy and gives measurements.
http://lwn.net/images/conf/rtlws11/papers/paper.37.html

.I The maxwell(8) random number generator
by Sandy Harris documents the design rationale for
this program.

That paper is included in ODF and PDF formats in the
source directory for the daemon. License is CC-by-SA;
the program and this man page are under GPL. If you
need either under some other open source license,
email me: sandyinchina@gmail.com.

src/maxwell.c
/*
	A daemon to get entropy from timer randomness
	and feed it to random(4)

	It borrows some code from Folkert van Heusden's
	timer entropy daemon http://www.vanheusden.com/te/
	but the two programs are quite different.

	License is GPL v2, the same as the earlier code.

	If anyone needs it under another Open Source
	license, they can contact me.

	This is version 2, 2016
	The main changes from v1, 2012, are:
		different command-line options
		constants LO_CLAIM & HI_CLAIM
		(instead of a different prime for each option)
		at startup, do fast output to fill up the pool
		(David Jasa's -G option is now the default)
		moving actual output from main() to do1k()
		random choice of delay as argument to do1k()

	Sandy Harris, sandyinchina@gmail.com
*/

#include "maxwell.h"
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#include <syslog.h>

#include <linux/random.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/stat.h>

int good_num(char *) ;

void usage() ;

/* minimum number of loops */
#define MIN_LOOPS 3

/* entropy claims */
#define LO_CLAIM 16
#define HI_CLAIM 30

/*
	global variables
	set in main()
	used by do1k(), log_error() and error_exit()
*/
char *prog_name ;
int user, demon, output, loops, mix, mul ;

/*
	primes taken from
	http://primes.utm.edu/lists/small/1000.txt

	used as microsecond delay values
*/
u32 primes[] = {89, 97, 101, 103, 107} ;	// median 101, mean 99.4
u32 smallp[] = {43, 47,  53,  57,  59} ;	// median  53, mean 51.8

/*
	constants borrowed from SHA-1
	reference used, RFC 3174

	used as initialisation constants
	for the entropy-collecting variable
*/
u32 sha_c[] = {0x67452301, 0xEFCDAB89, 0x98BADCFE, 0x10325476, 0xC3D2E1F0} ;

// forward declarations
void do1k( u32, u32 ) ;
void error_exit(const char *) ;
void message(const char *) ;

int main( int argc, char **argv)
{
	int i, p, fast, halt ;
	char *u, *v ;

	/*
		defaults
		might be overidden by options
	*/
	p = 0 ;			// no paranioa, three loops
	halt = 0 ;		// loop forever
	fast = 4 ;		// 4 K bits done quickly

	/*
		defaults
		always apply except in testing
	*/
	demon = 1 ;
	mix = 1 ;
	mul = MUL ;
	loops = MIN_LOOPS ;

	prog_name = *argv ;

	// documented in getuid(1) as "always successful"
	// so do not bother checking for failure
	user = geteuid() ;

	// check that the required clock is present
	// provided by real time library
	// got_clock() is in timermod.c with other clock stuff
	if( !got_clock() )
		error_exit( "cannot read real time clock" ) ;

	// process command-line arguments
	for( argc--, argv++ ; (argc > 0) && (*argv != NULL) ; argc--, argv++ )	{
		u = *argv ;
		if( u[0] != '-' )
			usage() ;
		switch( u[1] )	{
			case '0' :
				message( "-0 option asks for 0 output, makes no sense") ;
				usage() ;
			// any digit but zero
			case '1': case '2': case '3':
			case '4': case '5': case '6':
			case '7': case '8': case '9':
				v = u + 1 ;
				loops = atoi(v) ;
				// no mixing for test output
				mix = 0 ;
				mul = 1 ;
				// output to stdout
				demon = 0 ;
				break ;
			// output to stdout
			case 's':
				demon = 0 ;
				break ;
			// paranoia level
			case 'p' :
				v = argv[1] ;
				if( (v == NULL) || !good_num(v) )
					usage() ;
				p = atoi(v) ;
				// extra increment to skip over numeric argument
				argc-- ; argv++ ;
				break ;
			// initial fast data in K bits
			case 'f' :
				v = argv[1] ;
				if( (v == NULL) || !good_num(v) )
					usage() ;
				fast = atoi(v) ;
				argc-- ; argv++ ;
				break ;
			// total output in K bits, then halt
			case 'h' :
				v = argv[1] ;
				if( (v == NULL) || !good_num(v) )
					usage() ;
				halt = atoi(v) ;
				argc-- ; argv++ ;
				break ;
			default:
				usage() ;
		}
	}

	// not testing and not root: fail
	if( demon && (user != 0) )	{
		fprintf(stderr, "%s: must be root to run maxwell(8)\n", prog_name) ;
		fprintf(stderr, "%s: updates kernel entropy count, so root only\n", prog_name) ;
		fprintf(stderr, "%s: exception is 'maxwell -s' for testing\n", prog_name) ;
		exit(1) ;
	}

	// not testing and root
	if( demon )	{
		if( (output=open("/dev/random", O_WRONLY)) == -1)
			error_exit("failed to open /dev/random") ;
		if( daemon(-1, -1) == -1)
			error_exit("failed to become daemon process") ;
		openlog( prog_name, (LOG_CONS|LOG_PID), LOG_DAEMON) ;
	}
	// output to standard out
	else	output = 1 ;

	message( "maxwell(8) v 2" );

	// fprintf( stderr, "fast loop: p %d loops %d halt %d fast %d output %d demon %d\n", p, loops, halt, fast, output, demon ) ;

	// first loop, fast output to fill random(4) pool
	// uses smaller prime and default loops == MIN_LOOPS for speed
	// -p option has no effect here; not needed when claiming only LO_CLAIM
	// halt == 0 means run forever
	for( i = 0 ; (i < fast) && ((halt == 0) || (i < halt)) ; i++ )
		do1k( smallp[t5()], LO_CLAIM ) ;

	// -p takes effect here
	loops = MIN_LOOPS + 2*p ;
	// fprintf( stderr, "main loop: p %d loops %d halt %d fast %d output %d demon %d\n", p, loops, halt, fast, output, demon ) ;

	// main loop, halt == 0 means run forever
	for( /* continue with same i */ ; (halt == 0) || (i < halt) ; i++ )
		do1k( primes[t5()], HI_CLAIM ) ;

	message( "shutting down" );
	exit(0) ;
}

/*
	produce one K bit of output
	32 32-bit words
*/
void do1k( u32 delay, u32 claim )
{
	int i, j, k, ret ;
	u32 a, b ;

	// fprintf( stderr, "do1k delay %d claim %d mix %d loops %d\n", delay, claim, mix, loops) ;

	for( i = 0 ; i < 32 ; i++ )	{
		if( mix )
			// initialise with a somewhat random constant
			a = sha_c[t5()] ;
		else
			a = 0 ;
		// main loop; sample & mix
		for( j = 0 ; j < loops ; j++)	{
			// get 16 samples for entropy
			for( k = 0 ; k < 16 ; k++ )	{
				usleep(delay) ;
				// mix in a sample
				// multiplication spreads bits out
				a += ( t31() * mul ) ;
				// rotate left by two bits
				b = a >> 30 ;
				a = (a << 2) | b ;
			}
			if( loops > 1 )	{
				// rotate by one bit between loops
				b = a >> 31 ;
				a = (a << 1) | b ;
			}
		}
		// mix thoroughly
		if( mix )
			a = qht(a) ;
		// output 32 bits
		if( (ret = write( output, &a, 4)) != 4)
			error_exit("write to output file failed") ;
		// update entropy estimate
		if( demon )	{
			// a failure here is logged
			// but does not stop the program
			// it does no real harm
			if( (ioctl(output, RNDADDTOENTCNT, claim)) == -1)
				message("ioctl() failed") ;
		}
	}
}

void error_exit(const char *reason)
{
	if( demon )		// running as root
		syslog(LOG_ERR, "%s", reason) ;
	else
		fprintf(stderr, "%s: error: %s\n", prog_name, reason) ;
	exit(-1) ;
}

void message(const char *reason)
{
	if( demon )
		syslog(LOG_INFO, "%s", reason) ;
	// else
	// do nothing
}

void usage()
{
	fprintf(stderr, "usage: %s [-<digit>] [-s] [-p <number>] [-h <number>] [-f <number>]\n", prog_name) ;
	fprintf(stderr, "usage: %s <digit> must be in range 1-9, <number> 0-99\n", prog_name) ;
	exit(-1) ;
}

/*
	allows numbers 0 - 99, no + - or decimal point
*/
int good_num(char *p)
{
	int i ;
	for( i = 0 ; *p ; p++, i++ )	{
		if( isdigit(*p)	)
			// do nothing
			 ;
		else
			return(0) ;
	}
	return( (i>0) && (i<=2) ) ;
}